Localization support

jasonyess opened the pull request:
A wip implementation of site localization based on [Project Fluent](https://projectfluent.org)
The rendered language depends on the user's `locale` cookie, with English being the default

## License Acceptance

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

stadust commented:
Nice!!

stadust commented on pointercrate-core-pages/static/js/localization.js:
```diff
@@ -0,0 +1,54 @@
+
+        this.languages = Array.from(ul);
+
+        // hide the currently active language from the list
+        Array.from(ul)
+            .map(language => {
+                if (language.dataset.lang == document.documentElement.lang) {
+                    language.parentNode.style = "display: none";
+                } else {
+                    language.parentNode.style = "display: block";
+                    this.addSelectionListener(language.parentNode);
+                }
+            });
```
Why can't we do this on the backend? :o

jasonyess replied:
added with bf775e7, along with a `LocalizationConfiguration` state

stadust commented on pointercrate-core/src/localization.rs:
```diff
@@ -0,0 +1,36 @@
+    LOCALES.lookup(lang, text_id)
+}
+
+pub fn ftr<'a>(lang: &'static LanguageIdentifier, text_id: &str, args: &Vec<(&'static str, impl Into<FluentValue<'a>> + Clone)>) -> String {
+    let mut args_map: HashMap<Cow<'static, str>, FluentValue<'_>> = HashMap::new();
+
+    for arg in args {
+        args_map.insert(Cow::Borrowed(arg.0), arg.1.clone().into());
+    }
+
+    LOCALES.lookup_with_args(lang, text_id, &args_map)
+}
```
Maybe this would be more ergonomic as a macro?

jasonyess replied:
added with 7a9fe43, renamed from `ftr` to `trp` (parameterized translation) to better reflect fluent's terminology, i prefer the macro awesome change

stadust commented on pointercrate-demonlist-pages/src/account/demons.rs:
```diff
@@ -128,267 +133,268 @@ impl AccountPageTab for DemonsTab {
 }
 
-fn change_name_dialog() -> Markup {
+fn change_name_dialog(lang: &'static LanguageIdentifier) -> Markup {
```
I wonder if we could stuff the current language into a task local variable (https://docs.rs/tokio/latest/tokio/task/struct.LocalKey.html) to avoid having to pass it everywhere. Potentially not worth it though

jasonyess replied:
something like this is exactly what i was looking for a month ago, will push an implementation sometime soon (along with parameterized translation as a macro)

jasonyess replied:
added with 7a9fe43, seems like a good change for the most part

stadust commented on pointercrate-demonlist-pages/src/account/list_integration.rs:
```diff
@@ -167,32 +167,37 @@ impl AccountPageTab for ListIntegrationTab {
+                        (tr(lang, "claim-info-panel.info-a"))
                         br;
-                        "To initiate a claim, click the pen left of the 'Claimed Player' heading. Once initiated, you have an unverified claim on a player. These claims will then be manually verified by members of the pointercrate team. You can request verification in " a.link href=(self.0) {"this discord server"} "."
+                        (PreEscaped(ftr(lang, "claim-info-panel.info-b", &vec![
+                            (
+                                "discord",
+                                format!("<a class=\"link\" href=\"{}\">{}</a>", &self.0, tr(lang, "claim-info-panel.info-discord"))
+                            )
+                        ])))
```
uhhh, maybe we can just split this into 3 localized strings? pre-link, link and post-link?

jasonyess replied:
im not really a fan of this idea tbh, i would rather remove the discord field from `ListIntegrationTab` and move this link html to locales/core.ftl (along with the discord link), which i did with some other link elements

stadust replied:
mh, no, I don't think we should put any HTML into the localization files. And we also cannot hardcode these links because they are wrong for pointercrate forks. What is your concern with just splitting it into multiple localization keys?

jasonyess replied:
the reason why i prefer the the current approach is because it allows the translations to be written nicely like so:
`.redirect = Already have a pointercrate account? { -redirect-login(text: "Sign in") } instead.`

thinking about it some more i think splitting the keys and removing the html is fine, but in specific cases i could see it raising a mess
![image](https://github.com/user-attachments/assets/17854cf4-df63-4adb-a497-ea0cc6efaf90)
implementing the styling here would mean there would have to be many keys

im also not too fond with how the translations would turn out for large bodies like these, despite them still only requiring three keys:
![image](https://github.com/user-attachments/assets/e029e7c3-5968-46bc-951a-d6d3fd72b8fe)
![image](https://github.com/user-attachments/assets/2901adb6-66bf-4788-92f0-892088fea22f)

stadust replied:
> because it allows the translations to be written nicely

true, but it makes the rust side of things a mess :P Whereas just reading 3 or more consecutive keys as a single paragraph in the localization file is fairly easy to understand.

So I think my vote will be splitting it into multiple segments in the localization files

jasonyess replied:
yea i can work with that

how should we handle this though, should we drop the styling? it will probably make a big mess on both sides lol
![image](https://github.com/user-attachments/assets/f5aebfe5-b708-48e5-838d-5bcf290af089)

stadust replied:
That'd just be 6 strings, right? Unless enumerations work completely differently in other languages (which, tbh, they might, I'm not a linguist)

jasonyess replied:
i dont think grammar rules are a problem in this particular case, but for english the keys would be:

- "... whether the record is "
- ", "
- ", "
- ", "
- " or "
- "."

so yeah 6
regardless this is a lot messier than i think is necessary, i'll look into these problems in more detail once i find a solution for frontend translations

jasonyess replied:
made some changes with 21565a7

- there is no more html inside the localization files
- as opposed to creating multiple keys to be concatenated together, i opted with having the text that is being styled be in a separate key, where it is wrapped in whatever styling using the `html!` macro, and placed inside the original body of text with the `trp!` (parameterized translation) macro. with this approach, different languages having different grammar rules should not be a problem at all, and on the rust side the html should be a lot more readable

stadust commented:
aside: have you thought about localization for error messages? :o

jasonyess commented:
if you're talking about error pages, then yes, im planning on it
if you mean those popups like `Frontend Error, please contact list admin` or `Cloudflare rate limited` and stuff then also yes

stadust commented:
> if you're talking about error pages, then yes, im planning on it, though they might be super tricky if you mean those popups like `Frontend Error, please contact list admin` or `Cloudflare rate limited` and stuff then also yes

Ahhh, I actually forgot that we do some error messages on the frontend too. Admittedly, maybe we can just ignore them for now? Are there actually many strings constructed on the frontend? If its not many, we can also just like, hardcode them into a dictionary in a `<script>` block that gets injected into every page's head, lol. Ugly, but that's what's already done with the list size lol

jasonyess commented:
that's actually another idea i had to get translations to the frontend, but ive been kind of avoiding it in the hopes of finding a nicer and more flexible solution lol (though loading translations from the javascript isn't fleshing out very well atm from my current testing, so ill probably end up landing on something along those lines anyways)

there are a lot of english strings in the frontend code, like input validation messages, paginator list items in the managers in the user area, dialog confirmation messages, the list goes on, so i dont think ignoring them is an option 🥲

jasonyess commented:
7aab29c adds the ability to load fluent (.ftl) resources via `GET /static/core/ftl/<resource>/<uri..>` where `resource` represents the translation file name to retrieve, and `uri` represents the document location which this was sent from (im not sure if this one is necessary). this endpoint uses cookies to determine the correct language to respond with

- added a `loadResource()` method to the js which returns a Promise, so in the demons page for example, the demon addition form and demons paginator can be initialized _after_ the translations are loaded
- just like in the rust, i added `tr` and `trp` utility functions for loading translations, with the only differences lying in the syntax
- when initializing a `Page` struct, a vec of necessary resources must be passed (like "demon" and "form", representing the translation file names)
these are loaded into a `<script>` tag in the HTML head, where the js will load all of them and fire my `fluentresourcesloaded` event, which im still not sure would be of much use but seems like it could

i also reworked `LocalizationConfiguration` a little bit in 5b61075

jasonyess commented:
with 2c90e85 i created a new `pointercrate-core-macros` crate (i hope you're okay with this) and added two proc macros which, in the request handlers for each endpoint with a localization function nested somewhere inside it, eliminate the need to manually extract the client's locale and wrap the function body inside a `tokio::task_local!` `LocalKey` scope

this macro is used on any endpoint which returns a page or potentially returns an enum that implements `PointercrateError`

also added localization support for `CoreError` and `DemonlistError`

stadust commented on pointercrate-core-api/src/localization.rs:
```diff
@@ -0,0 +1,71 @@
+//
+// <resource> refers to the translation file name
+// <uri> refers to the uri that requested this resource
+#[rocket::get("/ftl/<resource>/<uri..>")]
```
(and cloudflare will automatically cache the localization files that way, which is a huge plus)

jasonyess commented on locales/en/documentation.ftl:
this is just for the topics in the table of contents in the api docs and the guidelines pages
i have the topic names here since i needed them in the localization setup for the different guideline languages (and since the docs and guidelines are basically identical functionally, i included the documentation topic names too for convenience; they don't actually need to be translated 🙂)
also i named it documentation.ftl since "documentation" is what seemed to be the closest to a "code name" for both lol

jasonyess commented on pointercrate-core-pages/static/js/modules/localization.js:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    let resourcePromises = [];
+    window.ftlResources.forEach((resource) => {
```
resolved with 0cb3051
definitely cleans things up a bit

jasonyess commented on pointercrate-core-api/src/maintenance.rs:
```diff
@@ -50,6 +50,8 @@ impl Fairing for MaintenanceFairing {
 #[rocket::get("/maintenance")]
-async fn maintenance() -> Result<()> {
-    Err(CoreError::ReadOnlyMaintenance.into())
+async fn maintenance(locale: ClientLocale) -> Result<()> {
```
yeahh the problem is the macro references `ClientLocale` as `pointercrate_core_api::localization::ClientLocale` for convenience in other crates, but this maintenance req handler is inside pointercrate_core_api, and i didn't think it was worth coming up with a solution for this since it's the only case of this happening

stadust commented on pointercrate-core-macros/src/lib.rs:
```diff
@@ -0,0 +1,65 @@
+
+            let __locale = match pointercrate_core_api::localization::ClientLocale::from_request(__request).await {
+                rocket::request::Outcome::Success(locale) => locale,
+                _ => todo!("this needs to be handled properly"),
```
just do internal server error here, we don't necessarily wanna panic in request handlers

jasonyess replied:
gotcha, that makes things a little easier haha

also yes a string is always returned from any call to `tr` or `trp`, and it'll default to "Unknown localization <text id>" if either it doesn't exist or it's given invalid parameters

stadust commented on pointercrate-core-pages/src/localization.rs:
```diff
@@ -0,0 +1,194 @@
+/// Withholds the site's core localization information.
+pub struct LocalizationConfiguration {
+    default: LocaleSet,
+    overrides: HashMap<PathBuf, LocaleSet>,
```
I'm on the fence as to whether we really need this, but I don't really think so. Let's go with the simpler option of "same language everywhere" for now

jasonyess replied:
sorry but im not really following the rationale of "same language everywhere" 😕

my reasoning for implementing this state was to allow different pages to have different sets of languages that they can be translated in, which would be most applicable in pointercrate.com's case for the api docs and the rest of the site (the api docs are only ever going to be in english, but that won't stop the rest of the site from being translated in like 3 other languages. having additional translations for just the guidelines would also be no problem)

also i decided on it being a state to allow clones to follow a different set of these "rules"

stadust replied:
Ahhhh, okay I understand a bit more now. I was missing that the docs page would switch the language selector back to English. I don't think we actually need to do that tbh. I was thinking that we'd just translate "API Documentation" as "API Documentation (English)" in all other languages (eg if the language is set to English, it'd just say "api documentation", but if the language is set to German it'd say "API Dokumentation (Englisch)")

jasonyess replied:
aaa sorry, that's actually an oversight on my end, ill definitely need to slightly rework the system (and the config struct) to achieve this

also sorry for leaving this so inactive for a while 😭😭

stadust commented on pointercrate-core-pages/static/js/modules/form.js:
```diff
@@ -1232,6 +1234,7 @@ export function displayError(output, specialCodes = {}) {
       output.setError(
         "FrontEnd JavaScript Error. Please notify an administrator and tell them as accurately as possible how to replicate this bug!"
       );
+      console.error(response);
```
ay, this'll fix #247 I guess

stadust replied:
nooo please leave it in, its useful haha

stadust commented on locales/en/home.ftl:
These need to be in the private repository. But I don't think that's a problem at all anymore now that ftl files are just normal static files?

Actually, can we move the locales for each part of the website into the respective crate? E.g. statsviewer.ftl goes into pointercrate-demonlist-pages/static/locales, and so on

jasonyess replied:
makes sense, ill just have to slightly rework the way the translation files are loaded at compile time

stadust commented on pointercrate-core/src/localization.rs:
```diff
@@ -0,0 +1,69 @@
+/// are stored in the `locales` directory.
+///
+/// This function call must be nested inside of a [`LocalKey`] scope.
+pub fn tr(text_id: &str) -> String {
+    LANGUAGE.with(|lang| LOCALES.lookup(lang, text_id))
+}
```
definitely doesnt need to happen in this PR, but it'd be really cool if these were proc-macros that (a) cause compilation errors if `text_id` does not exist in any locale, and (b) print warnings if some locale is missing `text_id`.

jasonyess replied:
yeah there are definitely ways to improve the system, including the things you mentioned, and i intend to come back to this later on and implement some additional things that would make using these functions a lot easier and safer

jasonyess commented on pointercrate-core-api/src/preferences.rs:
```diff
@@ -0,0 +1,132 @@
+}
+
+/// A request guard which stores the preferences sent from the client.
+pub struct ClientPreferences(Vec<ClientPreference>);
```
probably a better design choice than what i have haha

jasonyess commented on pointercrate-core-api/src/response.rs:
```diff
@@ -30,20 +38,41 @@ impl HeadLike for Page {
+            .by_code(preferences.get::<String>(locale_set.cookie))
+            .ok_or(Status::BadRequest)?;
+
+        let page_config = futures::executor::block_on(async {
```
yeah it's ugly but i settled on it because i don't think it has potential to cause major performance bottlenecks

i could also make it so the page config for every language is loaded in advance and passed around as a state, so it wouldn't have to go through this
now that i think about it this approach would probably be more optimal

stadust replied:
Mhhh that'd work, or just having the function take a locale argument, and bringing back variants of tr/trp! that do not look at the task local for use by it. But like I said, can also stay the way it is. I do agree that it's not likely to cause problems. At least I don't see how it could haha

jasonyess commented:
with 441b70b
completely rewrote the way translations are loaded and dropped the use of the fluent_templates crate so

- required translations can be loaded in binary crates, useful because pointercrate-example and pointercrate-bin should load a different set of fluent resources (like home.ftl shouldn't be here, and the nav.ftl files are different)
- different binary crates can also have different sets of languages now, so we don't need to specify that the guidelines page can be translated in russian here, since that's specific to pointercrate-bin
- allows us to have more control over handling errors and stuff
- can now have the ftl files go into their specific crates 🙂

also added some better error handling on the rust side of the localization system so it shouldn't panic at all in pretty much all cases

stadust commented on pointercrate-core/src/localization.rs:
```diff
@@ -0,0 +1,148 @@
+
+    /// Set the `LOCALES` [`OnceLock`] to use this set of loaded locales
+    pub fn commit(self) {
+        LOCALES.set(self).unwrap_or_else(|_| panic!("Failed to set LOCALES OnceLock"));
```
`.expect()` here

stadust commented on pointercrate-example/src/main.rs:
```diff
@@ -48,11 +55,31 @@ fn home() -> Redirect {
     // Load the configuration from your .env file
     dotenv::dotenv().unwrap();
 
+    LocalesLoader::new(&SUPPORTED_LOCALES[0])
+        .locale(
+            &SUPPORTED_LOCALES[0],
+            vec![
+                "pointercrate-core-pages/static/ftl/en/error.ftl",
+                "pointercrate-core-pages/static/ftl/en/nav.ftl",
+                "pointercrate-core-pages/static/ftl/en/ui.ftl",
+                "pointercrate-demonlist-pages/static/ftl/en/demon.ftl",
+                "pointercrate-demonlist-pages/static/ftl/en/overview.ftl",
+                "pointercrate-demonlist-pages/static/ftl/en/player.ftl",
+                "pointercrate-demonlist-pages/static/ftl/en/record.ftl",
+                "pointercrate-demonlist-pages/static/ftl/en/statsviewer.ftl",
+                "pointercrate-demonlist-pages/static/ftl/en/submitter.ftl",
+                "pointercrate-user-pages/static/ftl/en/user.ftl",
+            ],
+        )
+        .commit();
```
Let's do this slightly differently: `LocalLoader` just take a vec of `ftl` directories (e.g. in this case, `pointercrate-core-pages/static/ftl`, `pointercrate-demonlist-pages/static/ftl` and `pointercrate-user-pages/static/ftl`), and then dynamically determine the available locales from the child-directories (e.g. in this case, just `en/`), and then just traverse one layer lower to find all actual `.ftl` files. It'd be neat if it could even print warnings if the available locales or `.ftl` files are inconsistent across static directories / locales respectively.

jasonyess replied:
added with adf2e57 (and cdd5ede i guess, no warning system in either though 😞)
`LocalesLoader::load` takes in /static directories and *only* searches the ones whose name exists in the given list of supported languages

i don't really think determining the available locales through the static directories is better in this case, since not all directories are consistent with the languages they contain (e.g. bin repo's /static has en and ru while everything else only has en), and determining the site urls where the languages apply is another problem which i really didnt want to deal with 😭😭 since thats already handled in `LocalizationConfiguration`

also used `add_resource_overriding` since the nav.ftl files in pointercrate and pointercrate-bin differ

also removed fallback logic from `LocalesLoader` since it's already handled with `LocalizationConfiguration` when receiving requests (a3e579e)

stadust commented on pointercrate-core/build.rs:
```diff
@@ -1,5 +1,7 @@
     // trigger recompilation when a new migration is added
+    // or a new translation file is added
     println!("cargo:rerun-if-changed=../migrations");
+    println!("cargo:rerun-if-changed=../locales");
```
not needed anymore I guess?

stadust commented:
One last request, could you reorganize the commit history a bit? E.g. one commit that adds all of the framework and stuff, and then another commit that does all the conversion of the existing strings? Right now its a bit hard to review the details of the framework because its drowned in all the diff from the conversion :/

stadust commented:
Thanks, looks good to me now! I'll fix up any remaining small things myself when I merge this. I'll try to collect together the framework pieces and push them to `master` next week :)

jasonyess commented:
sounds good!! i went over everything and it seems to be in a complete state (minus potential convenience features)
also ill try reorganizing the history to make your life easier haha
